### PR TITLE
Property implicits cleanup

### LINKED ISSUE #271: Divergent Implicit Type Error for SeqProperty
The following code puts the compiler into an infinite loop possibly because of a DivergentImplicitTypeError (Not completely sure it is infinite but I gave it a solid 20 minutes before I killed it):
```scala
private case class Chunk[T](
    var requested: Boolean = false,
    data: SeqProperty[T] = SeqProperty.blank
)
```

If I change `SeqProperty.blank` to `SeqProperty.blank[T]` it will throw an error saying it cannot find a implicit ProperyCreator[T] as it should.

### PR DESCRIPTION
**Pull request by ddworak**
Closes #271 

Towards #272

### CODE REVIEW

File: core/src/main/scala/io/udash/properties/single/CastableProperty.scala [hunk 1]
```diff
@@ -18,14 +18,23 @@ trait CastableReadableProperty[A] extends ReadableProperty[A] {
...
-  */
+ * <b>NOTE:</b> Those casts are checked in compilation time: "it compiles" == "it works", as long as the property creators
+ * were not explicitly passed and relevant implicit scope is equal for property creation and cast.
+ */
 trait CastableProperty[A] extends CastableReadableProperty[A] with Property[A] {
-  /** Safely casts DirectProperty[A] to ModelProperty[A] */
-  override def asModel(implicit ev: ModelPropertyCreator[A]): ModelProperty[A] =
+  /** Safely casts `DirectProperty[A]` to `ModelProperty[A]` */
+  override def asModel(implicit ev: ModelPropertyCreator[A]): ModelProperty[A] = {
+    if (!this.isInstanceOf[ModelProperty[A]])
```
Comment (ghik) [inline core/src/main/scala/io/udash/properties/single/CastableProperty.scala]: pattern match?

File: core/src/main/scala/io/udash/properties/single/CastableProperty.scala [hunk 2]
```diff
@@ -18,14 +18,23 @@ trait CastableReadableProperty[A] extends ReadableProperty[A] {
...
+      throw new IllegalStateException("Property was created without provided ModelPropertyCreator in scope. " +
+        "Make sure it is uniformly available, e.g. in the companion object of the model class.")
     this.asInstanceOf[ModelProperty[A]]
+  }
 
-  /** Safely casts DirectProperty[Seq[A]] to DirectSeqProperty[A] */
-  override def asSeq[B](implicit sev: A <:< Seq[B], ev: SeqPropertyCreator[B]): SeqProperty[B, CastableProperty[B]] =
+  /** Safely casts `DirectProperty[Seq[A]]` to `DirectSeqProperty[A]` */
+  override def asSeq[B](implicit sev: A <:< Seq[B], ev: SeqPropertyCreator[B]): SeqProperty[B, CastableProperty[B]] = {
+    if (!this.isInstanceOf[SeqProperty[_, _]])
```
Comment (ghik) [inline core/src/main/scala/io/udash/properties/single/CastableProperty.scala]: pattern match?

File: macros/src/main/scala/io/udash/macros/PropertyMacros.scala [hunk 3]
```diff
@@ -217,7 +218,12 @@ class PropertyMacros(val ctx: blackbox.Context) extends AbstractMacroCommons(ctx
         members.map {
           case (name, returnTpe) =>
             val reifiedReturnTpe = reifySeqTpe(returnTpe)
-            q"""properties(${name.toString}) = implicitly[$PropertyCreatorCls[$reifiedReturnTpe]].newProperty(null.asInstanceOf[$reifiedReturnTpe], this)"""
+            val reifiedCreatorTpe = getType(
+              if (tpe <:< SeqTpe) tq"$SeqPropertyCreatorCls[${tpe.typeArgs.head}]"
```
Comment (ghik) [inline macros/src/main/scala/io/udash/macros/PropertyMacros.scala]: When accessing `typeArgs`, beware of:
* type aliases
* classes that extend `Seq` but don't have the same parameterization scheme, e.g. `class WeirdSeq[A,B] extends Seq[B]` or `class IntSeq extends Seq[Int]` or `class SeqSeq[T] extends Seq[Seq[T]]`

File: core/src/main/scala/io/udash/properties/PropertyCreatorImplicits.scala [hunk 4]
```diff
@@ -1,57 +1,9 @@
 package io.udash.properties
 
-trait PropertyCreatorImplicitsLow {
-  implicit def materializeSingle[T]: PropertyCreator[T] =
-    macro io.udash.macros.PropertyMacros.reifyPropertyCreator[T]
+trait PropertyCreatorImplicitsLow { this: PropertyCreatorImplicits =>
```
Comment (ghik) [inline core/src/main/scala/io/udash/properties/PropertyCreatorImplicits.scala]: `this: PropertyCreator.type =>` ?

File: core/src/main/scala/io/udash/properties/single/CastableProperty.scala [hunk 5]
```diff
@@ -24,17 +24,19 @@ trait CastableReadableProperty[A] extends ReadableProperty[A] {
...
-    this.asInstanceOf[ModelProperty[A]]
+    }
   }
 
   /** Safely casts `DirectProperty[Seq[A]]` to `DirectSeqProperty[A]` */
   override def asSeq[B](implicit sev: A <:< Seq[B], ev: SeqPropertyCreator[B]): SeqProperty[B, CastableProperty[B]] = {
-    if (!this.isInstanceOf[SeqProperty[_, _]])
-      throw new IllegalStateException("Property was created without provided SeqPropertyCreator in scope. " +
+    this match {
+      case sp: this.type with SeqProperty[_, _] => sp.asInstanceOf[SeqProperty[B, CastableProperty[B]]]
```
Comment (ghik) [inline core/src/main/scala/io/udash/properties/single/CastableProperty.scala]: Why do you need this `this`?

File: core/src/main/scala/io/udash/properties/single/CastableProperty.scala [hunk 6]
```diff
@@ -18,14 +18,25 @@ trait CastableReadableProperty[A] extends ReadableProperty[A] {
...
+    }
+  }
 
-  /** Safely casts DirectProperty[Seq[A]] to DirectSeqProperty[A] */
-  override def asSeq[B](implicit sev: A <:< Seq[B], ev: SeqPropertyCreator[B]): SeqProperty[B, CastableProperty[B]] =
-    this.asInstanceOf[SeqProperty[B, CastableProperty[B]]]
+  /** Safely casts `DirectProperty[Seq[A]]` to `DirectSeqProperty[A]` */
+  override def asSeq[B](implicit sev: A <:< Seq[B], ev: SeqPropertyCreator[B]): SeqProperty[B, CastableProperty[B]] = {
+    this match {
+      case sp: this.type with SeqProperty[_, _] => sp.asInstanceOf[SeqProperty[B, CastableProperty[B]]]
```
Comment (ghik) [inline core/src/main/scala/io/udash/properties/single/CastableProperty.scala]: Why do you need this `this`?

Comment (ddworak) [reply core/src/main/scala/io/udash/properties/single/CastableProperty.scala]: I wish I knew:
```scala
Error:(37, 16) pattern type is incompatible with expected type;
 found   : io.udash.properties.seq.SeqProperty[(some other)_(in method asSeq),_(in method asSeq)]
 required: io.udash.properties.single.CastableProperty[A]
      case sp: SeqProperty[_, _] => sp.asInstanceOf[SeqProperty[B, CastableProperty[B]]]
```

**Pull request merged**